### PR TITLE
A powerful API for endless scrolling

### PR DESCRIPTION
**Pull request by jasonsparc**
I would like to add `EndlessScrollHelper`
- An extension of `EndlessRecyclerOnScrollListener` that allows a more granular implementation of `onLoadMore()`.
- Providing an easy construct for running an asynchronous loader to load the new items and automatically handling the delivery of those stuff to the main thread.

It also comes with a utility that delivers the new items to an `IItemAdapter` or `GenericItemAdapter`.

I'm not sure if I'm convinced with the naming so feel free to criticize! ;)

### CODE REVIEW

**Comment by jasonsparc** [comment]
Following up with some more javadocs

**Comment by jasonsparc** [comment]
Okay! I'm probably done adding new commits. :)

**Comment by mikepenz** [comment]
@jasonsparc oh cool. Will have a closer look later today. But looks really awesome ;). Thanks

**Comment by mikepenz** [comment]
Might be helpful to have a sample showcasing the usage of the helper.

**Comment by jasonsparc** [comment]
@mikepenz Sorry if I wasn't able to provide a sample. I'll try to push a sample tomorrow :D

Currently, I'm using it like this:

```java
private void setUpFastAdapter(Bundle savedInstanceState) {
	final ItemAdapter itemAdapter = new ItemAdapter();
	itemAdapter.wrap(fastAdapter);

	final FooterAdapter loaderFooter = new FooterAdapter();
	loaderFooter.wrap(fastAdapter);

	...

	List<ProgressItem> loaderFooterItems = Collections.singletonList(new ProgressItem());
	loaderFooter.setNewList(loaderFooterItems);

	// TODO optimize loadMoreThreshold parameter according to screen size
	final int loadMoreThreshold = 30;
	// TODO optimize size parameter according to largest screen size
	final int size = 7; // request in small bursts

	StaticState state = this.state.get();

	RecyclerView recycler = binding.recycler;
	new EndlessScrollHelper<Project>(recycler.getLayoutManager(), loadMoreThreshold)
			.withOnLoadMoreHandler((out, currentPage) -> {
				currentPage = currentPage - 1; // Sorry, we use zero-based indexes
				ZipMatch.api().getBookmarkedProjects(currentPage, size)
						.subscribeOn(Sched.io())
						.observeOn(Sched.computation())
						.compose(Op.catchError(state.errors))
						.observeOn(Sched.main())
						.subscribe(out::deliverNewItems);
			})
			.withNewItemsDeliveredTo(itemAdapter, this::toItem, (newItems, page) -> {
				if (newItems.isEmpty() || newItems.size() < size) {
					// An empty return or a return with less items than expected indicates we have just finished loading.
					loaderFooter.setNewList(Collections.emptyList());
				} else {
					this.items.addAll(newItems);
				}
			})
			.addTo(recycler)
			.onLoadMore(1); // Perform initial loading
}
```

**Comment by mikepenz** [comment]
No problem.

File: library-extensions/src/main/java/com/mikepenz/fastadapter_extensions/scroll/EndlessRecyclerOnScrollListener.java [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
     private int mCurrentPage = 1;
 
-    private RecyclerView.LayoutManager mLayoutManager;
+    RecyclerView.LayoutManager mLayoutManager;
```
Comment (FabianTerhorst) [inline library-extensions/src/main/java/com/mikepenz/fastadapter_extensions/scroll/EndlessRecyclerOnScrollListener.java]: you should use the getter for the layout manager #98

Comment (jasonsparc) [inline library-extensions/src/main/java/com/mikepenz/fastadapter_extensions/scroll/EndlessRecyclerOnScrollListener.java]: @FabianTerhorst Cool! :) I'll be pushing new commits.

**Pull request merged**